src/main/java/com/squareup/kotlinpoet/TypeSpec.kt

```diff
@@ -519,6 +530,24 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
         "anonymous type has too many supertypes"
       }
 
+      if (superinterfaceDelegates.isNotEmpty()) {
+        requireNotNull(primaryConstructor) {
+          "delegate superinterfaces ${superinterfaceDelegates.map { it.type }} " +
+              "require value parameters ${superinterfaceDelegates.map { it.name }} by primary constructor" }
+
+        superinterfaceDelegates.forEach {
+          require(!superinterfaces.contains(it.type)) {
+            "class '$name' already implements ${it.type}, delegating to constructor parameter '${it.name}' not allowed" }
+        }
+
+        if(primaryConstructor?.parameters?.containsAll(superinterfaceDelegates) == false)
+            primaryConstructor = primaryConstructor!!.toBuilder()
+                .addParameters(superinterfaceDelegates)
+                .build()
+        // re-assigning ^^^ allowed?? otherwise need to require non-null ctor at time of declaring delegate
```

Comment:
This validation can be made in `build()`, when we're sure all the args have been provided